toptal/picasso
# feat(DatePicker): added ability to pass custom parsers for DatePicker's Input

[FX-2324]

### Description

To implement human-readable date parsing for DatePicker in Staff-Portal we need to introduce a way to pass custom parsers for DatePicker's Input value. We should apply a custom parser on `onChange` event of DatePicker’s Input.

To have more details, please read all spike comments https://toptal-core.atlassian.net/browse/SPT-216

### How to test

- Open [Story Book](https://picasso.toptal.net/FX-2324-custom-date-picker-parsers/?path=/story/picasso-lab-datepicker--datepicker)
- Scroll down to the last example
- This DatePicker has a custom parser that parses the input value containing the year only (`2015`, `015`, `15`) and returns a random date within this year
- Selected day in the calendar should be updated when changing input value (`2015` => `2014`)
- Previous valid date should be selected when entering an invalid value (`abc`)

### Screenshots

### Review


<details>
<summary>PR commands</summary>
<br />

List of available commands:

- `@toptal-bot run all` - Run whole pipeline
- `@toptal-bot run build` - Check build
- `@toptal-bot run visual` - Run visual tests
- `@toptal-bot run deploy:documentation` - Deploy documentation
- `@toptal-bot run package:alpha-release` - Release alpha version

</details>

<details>
<summary>PR Review Guidelines</summary>
<br />

#### When to approve? ✅

**You are OK** with merging this PR and
1. You have no extra requests.
2. You have optional requests.
   1. Add `nit: ` to your comment. (ex. `nit: I'd rename this variable from makeCircle to getCircle`)

#### When to request changes? ❌

**You are not OK** with merging this PR because
1. Something is broken after the changes.
2. Acceptance criteria is not reached.
3. Code is dirty.

#### When to comment (neither ✅ nor ❌)

**You want your comments to be addressed** before merging this PR in cases like:
1. There are leftovers like unnecessary logs, comments, etc.
2. You have an opinionated comment regarding the code that requires a discussion.
3. You have questions.

#### How to handle the comments?

1. An owner of a comment is the only one who can resolve it.
2. An owner of a comment must resolve it when it's addressed.
3. A PR owner must reply with ✅ when a comment is addressed.

</details>


[FX-2324]: https://toptal-core.atlassian.net/browse/FX-2324?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
